statsd/container.go

```diff
@@ -0,0 +1,59 @@
+package statsd
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+)
+
+const (
+	// cgroupPath is the path to the cgroup file where we can find the container id if one exists.
+	cgroupPath = "/proc/self/cgroup"
+)
+
+const (
+	uuidSource      = "[0-9a-f]{8}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{12}"
+	containerSource = "[0-9a-f]{64}"
+	taskSource      = "[0-9a-f]{32}-\\d+"
+)
+
+var (
+	// expLine matches a line in the /proc/self/cgroup file. It has a submatch for the last element (path), which contains the container ID.
+	expLine = regexp.MustCompile(`^\d+:[^:]*:(.+)$`)
+
+	// expContainerID matches contained IDs and sources. Source: https://github.com/Qard/container-info/blob/master/index.js
+	expContainerID = regexp.MustCompile(fmt.Sprintf(`(%s|%s|%s)(?:.scope)?$`, uuidSource, containerSource, taskSource))
+)
+
+// parseContainerID finds the first container ID reading from r and returns it.
+func parseContainerID(r io.Reader) string {
+	scn := bufio.NewScanner(r)
+	for scn.Scan() {
+		path := expLine.FindStringSubmatch(scn.Text())
+		if len(path) != 2 {
+			// invalid entry, continue
+			continue
+		}
+		if parts := expContainerID.FindStringSubmatch(path[1]); len(parts) == 2 {
+			return parts[1]
+		}
+	}
+	return ""
+}
+
+// readContainerID attempts to return the container ID from the provided file path or empty on failure.
+func readContainerID(fpath string) string {
+	f, err := os.Open(fpath)
+	if err != nil {
+		return ""
+	}
+	defer f.Close()
+	return parseContainerID(f)
+}
+
+// getContainerID attempts to return the container ID from /proc/self/cgroup or empty on failure.
+var getContainerID func() string = func() string {
+	return readContainerID(cgroupPath)
+}
```

Comment:
Since the containerID is never going to change, could we save it in a var at the package level and initialized it only once with `sync.Once`.

Then `getContainerID` will just return that variable. This would avoid passing the container ID everywhere in the code. We could just call `getContainerID` in `appendContainerID` (in the format.go file). We could then control the value of that variable for easy testing.

This would simplify the code by a lot and avoid passing the same string every time.

WDYT ?

Reply:
> Since the containerID is never going to change, could we save it in a var at the package level and initialized it only once with sync.Once.

👍 

> Then getContainerID will just return that variable. This would avoid passing the container ID everywhere in the code. We could just call getContainerID in appendContainerID (in the format.go file). We could then control the value of that variable for easy testing.
> This would simplify the code by a lot and avoid passing the same string every time.
> WDYT ?

Sounds good, the reason why I implemented it that way was because I wanted to keep it consistent with the existing code and how metric properties are passed to the buffering logic.

Reply:
Yes good point, so you're suggesting adding a `WithoutContainerID(id string)` option to the builder in addition to `WithoutOriginDetection`.

Reply:
Yes, except I think we should name them with a positive preposition: `WithContainerID` and `WithOriginDetection`.

Reply:
I tried to apply all your suggestions in the latest commit, lmk how it looks :)

re `WithOriginDetection` we're going to provide both with and without, see https://github.com/DataDog/datadog-go/pull/250#discussion_r795581889

Reply:
Yes we offer both `With` and `Without` in the client.